### DISCUSSION TITLE
Question about proxy examples for websocket

### DISCUSSION THREAD

**Discussion by jsbrain** [dani-garcia/vaultwarden#3321] [Q&A] [answered]
Almost all proxy examples have a rule to mach `/notifications/hub` but not `/notifications/hub/negotiate` as the latter is apparently being used for http.

The matching in most of the examples is done via "starts with" and I'm curious why. As I understand, the `/notifications/hub` is a fixed socket endpoint which does not have any other parts in the path url or am I wrong here?

Because if so, most of the examples could be simplified while some (pfSense example) are just plain wrong.

Clarification on that matter would be highly appreciated 🙏🏻

**Comment by stefan0xC** [comment]
I've only looked at the nginx example as that is the proxy I'm most familiar with and I think you're right that it could be simplified by  dropping the `location /notifications/hub/negotiate` block and matching `location = /notifications/hub` exactly. Some of the other examples (like `Caddy 2.x` and `lighttpd (by forkbomb9)`) seem to already use exact matching but I've not checked all of them.

**Reply by jsbrain** [reply]
Ok so my assumption is correct, the ws path `/notifications/hub` is always static and has no other parts involved, yes?

If so, I'm happy to go through the examples and provide corrections where I can. Especially the HAProxy examples I can handle for sure, probably the others as well.

**Reply by kpfleming** [reply]
That may be true today, but the Bitwarden clients could change in the future to use more sub-paths under `/notifications/hub` which are not `negotiate`, and when Vaultwarden was upgraded to provide support for them users would have to modify their proxy configurations.

If the current configuration isn't causing any harm, it may be better to leave it alone.

**Reply by jsbrain** [reply]
Your argument is against you. As soon as there's another http path, e.g. `/notifications/hub/new` the current configs would need to be changed.

Whereas if actually only routes are configured, that are actually there, like in pseudo:

```conf
if { /notifications/hub } route ws
else route http
```

users wouldn't have to change anything and the configuration still works. Only if there's changes to the ws endpoint the configs would need to be edited.

I'd argue it's generally a better approach to **not** configure something that's actually not there as this could end up being highly misleading and hard to debug.

**Reply by jsbrain** [reply]
I think it's better to explicitly configure the ws route only and else default to http as ws routes usually don't change that often and usually don't use any (sub-)path sections.

**Reply by jsbrain** [reply]
To illustrate:

If you take the example: HAproxy (by [@williamdes](https://github.com/williamdes))
```
backend static-success-default
  mode http
  errorfile 503 /usr/local/etc/haproxy/static/index.static.default.html
  errorfile 200 /usr/local/etc/haproxy/static/index.static.default.html

frontend http-in
    bind *:80
    bind *:443 ssl crt /acme.sh/domain.tld/domain.tld.pem
    option forwardfor header X-Real-IP
    http-request set-header X-Real-IP %[src]
    default_backend static-success-default

    # Define hosts
    acl host_bitwarden_domain_tld hdr(Host) -i bitwarden.domain.tld

    ## figure out which one to use
    use_backend vaultwarden_http if host_bitwarden_domain_tld !{ path_beg /notifications/hub } or { path_beg /notifications/hub/negotiate }
    use_backend vaultwarden_ws if host_bitwarden_domain_tld { path_beg /notifications/hub } !{ path_beg /notifications/hub/negotiate }

backend vaultwarden_http
    # Enable compression if you want
    # compression algo gzip
    # compression type text/plain text/css application/json application/javascript text/xml application/xml application/xml+rss text/javascript
    # You can use the container hostname if you are using haproxy with docker-compose
    server vw_http 0.0.0.0:8080

backend vaultwarden_ws
    # You can use the container hostname if you are using haproxy with docker-compose
    server vw_ws 0.0.0.0:3012
```


This:

```conf
## figure out which one to use
use_backend vaultwarden_http if host_bitwarden_domain_tld !{ path_beg /notifications/hub } or { path_beg /notifications/hub/negotiate }
use_backend vaultwarden_ws if host_bitwarden_domain_tld { path_beg /notifications/hub } !{ path_beg /notifications/hub/negotiate }
```

is literally the same as this:

```conf
## figure out which one to use
use_backend vaultwarden_ws if host_bitwarden_domain_tld { path_beg /notifications/hub }
use_backend vaultwarden_http if host_bitwarden_domain_tld
```

**Reply by jsbrain** [reply]
Another indicator that this might confuse some people (as it did confuse me at first) is the pfSense example.

Apparently @RichardMawdsley pretty much got confused while trying to convert the HAProxy example to pfSense (no offense) because his config can actually be reduced to

```conf
acl			ACL00	var(txn.txnhost) -m str -i VAULTWARDEN.MYDOMAIN.COM
acl			ACL1	var(txn.txnpath) -m str -i /notifications/hub

use_backend VaultWarden-Notifications_ipvANY  if  ACL00 ACL1
use_backend VaultWarden_ipvANY  if  ACL00
```

and just works flawlessly.

**Reply by stefan0xC** [reply]
> is literally the same as this:
>
>     ## figure out which one to use
>     use_backend vaultwarden_ws if host_bitwarden_domain_tld { path_beg /notifications/hub }
>     use_backend vaultwarden_http if host_bitwarden_domain_tld


I'm not that familiar with HAProxy but shouldn't it just be `{ path /notifications/hub }` so the matching is exact? Not sure about the existing config either.

**Reply by jsbrain** [reply]
You're right, that's what I meant ofc. Exact match!

**Reply by MawdleCode** [reply]
Wouldn't be suprised.  I was fumbling around until I found something that worked!

**Reply by jsbrain** [reply]
Just double checked, it's `{ path -m str /notifications/hub }` for strict match.

[Ref](https://www.haproxy.com/documentation/hapee/latest/configuration/acls/syntax/#:~:text=or%20string%20samples.-,str,-Exact%20string%20match)

**Reply by jsbrain** [reply]
> Wouldn't be suprised. I was fumbling around until I found something that worked!

Don't worry ... I can totally relate

**Comment by BlackDex** [comment]
I actually think that `negotiate` endpoint isn't used anymore. I think we can remove that whole part.
And only the `/notifications/hub` needs to be there to make it a web-socket, and that is about it.

**Reply by BlackDex** [reply]
I quickly removed that endpoint from the code, and all works just fine.
I have not tested the extension or desktop client yet btw.

**Reply by jsbrain** [reply]
~~Everything works fine for me, tested desktop and mobile client~~

Oh sorry you meant you removed the endpoint completely, I thought you were talking about the proxy config.

**Comment by jsbrain** [comment, marked as answer]
So this would lead us to:


HAproxy (by BlackDex):

```conf
frontend vaultwarden
    bind 0.0.0.0:80
    option forwardfor header X-Real-IP
    http-request set-header X-Real-IP %[src]
    use_backend vaultwarden_ws if { path -m str /notifications/hub }
    default_backend vaultwarden_http

backend vaultwarden_http
    # Enable compression if you want
    # compression algo gzip
    # compression type text/plain text/css application/json application/javascript text/xml application/xml application/xml+rss text/javascript
    server vwhttp 0.0.0.0:8080

backend vaultwarden_ws
    server vwws 0.0.0.0:3012
```


HAproxy (by @williamdes):

```conf
backend static-success-default
  mode http
  errorfile 503 /usr/local/etc/haproxy/static/index.static.default.html
  errorfile 200 /usr/local/etc/haproxy/static/index.static.default.html

frontend http-in
    bind *:80
    bind *:443 ssl crt /acme.sh/domain.tld/domain.tld.pem
    option forwardfor header X-Real-IP
    http-request set-header X-Real-IP %[src]
    default_backend static-success-default

    # Define hosts
    acl host_bitwarden_domain_tld hdr(Host) -i bitwarden.domain.tld

    ## figure out which one to use
    use_backend vaultwarden_ws if host_bitwarden_domain_tld { path -m str /notifications/hub }
    use_backend vaultwarden_http if host_bitwarden_domain_tld

backend vaultwarden_http
    # Enable compression if you want
    # compression algo gzip
    # compression type text/plain text/css application/json application/javascript text/xml application/xml application/xml+rss text/javascript
    # You can use the container hostname if you are using haproxy with docker-compose
    server vw_http 0.0.0.0:8080

backend vaultwarden_ws
    # You can use the container hostname if you are using haproxy with docker-compose
    server vw_ws 0.0.0.0:3012
```

HAproxy inside PfSense (by @RichardMawdsley):

Frontend:
<img width="1142" alt="Screen Shot 2023-03-08 at 13 26 09" src="https://user-images.githubusercontent.com/20225909/223713725-bf6da450-2a23-4879-aa0e-6ee1a024104f.png">

Backend:
<img width="1136" alt="Screen Shot 2023-03-08 at 13 28 23" src="https://user-images.githubusercontent.com/20225909/223713996-c4d216c8-0d93-4528-8be4-97eb0a112a54.png">
